### PR TITLE
add cloudio pickle patching for unified package

### PR DESCRIPTION
**Pull request by justusschock**
## What does this PR do?
Part of #15105 

In legacy versions of PL we pickled callbacks for checkpointing. Loading these checkpoints now results in importing PL even for the unified package (or failing if PL isn't present).

This unpickler patch redirects all imports from PL to the `lightning.pytorch` package if necessary.

Fixes #\<issue_number>

### Does your PR introduce any breaking changes? If yes, please list them.

None

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @borda

### CODE REVIEW

**Comment by justusschock** [comment]
should be ready for review now

**Comment by justusschock** [comment]
@carmocca the cleaning of sys modules we talked about caused a lot of other tests to crash. So this is the solution I came up with instead.

File: tests/tests_pytorch/graveyard/test_cloud_io.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+LEGACY_CHECKPOINTS_PATH = os.path.join(_PATH_LEGACY, "checkpoints")
+CHECKPOINT_EXTENSION = ".ckpt"
+# load list of all back compatible versions
+with open(os.path.join(_PROJECT_ROOT, "legacy", "back-compatible-versions.txt")) as fp:
+    LEGACY_BACK_COMPATIBLE_PL_VERSIONS = [ln.strip() for ln in fp.readlines()]
+
+
+@pytest.mark.parametrize("pl_version", LEGACY_BACK_COMPATIBLE_PL_VERSIONS)
+@pytest.mark.skipif(
+    not "pytorch_" + "lightning" in sys.modules, reason="This test is only relevant for the standalone package"
```
Comment (awaelchli) [inline tests/tests_pytorch/graveyard/test_cloud_io.py:21]: the string addition can be collapsed in a single string, right?

File: src/pytorch_lightning/_graveyard/cloud_io.py [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+        if module.startswith("pytorch_" + "lightning"):
+            # for the standalone package this won't do anything,
+            # for the unified mirror package it will redirect the imports
+
+            old_module = deepcopy(module)
+            module = "pytorch_lightning" + module[len("pytorch_" + "lightning") :]
+
+            # this warning won't trigger for standalone as these imports are identical
+            if module != old_module:
+                warnings.warn(f"Redirecting import of {module}.{name} to pytorch_lightning.{name}")
```
Comment (awaelchli) [inline src/pytorch_lightning/_graveyard/cloud_io.py:18]: What is the motivation for showing this to the user? Maybe we could say what they should do to avoid the warning?

**Review by awaelchli — APPROVED** [review]
Thanks looking into this probably very hard to debug issue! 
I hope we're not looking into the entrance of a rabbit hole 😅

File: src/pytorch_lightning/_graveyard/cloud_io.py (hunk 2 above)
Comment (Borda) [reply src/pytorch_lightning/_graveyard/cloud_io.py:18]: note that in a mono-package it will be replaced to be
```
warnings.warn(f"Redirecting import of {module}.{name} to lightning.pytorch.{name}")
```

File: tests/tests_pytorch/graveyard/test_cloud_io.py (hunk 1 above)
Comment (Borda) [reply tests/tests_pytorch/graveyard/test_cloud_io.py:21]: Yes and no, we want to preserve this even in the mono-package and the setup is running a naive replacement, so to prevent it we need to be creative...

**Review by Borda — APPROVED** [review]

File: tests/tests_pytorch/graveyard/test_cloud_io.py (hunk 1 above)
Comment (justusschock) [reply tests/tests_pytorch/graveyard/test_cloud_io.py:21]: It could, but then this would be replaced when converting the tests for the unified mirror package which we don't want. So this is kind of hacky to prevent this

File: src/pytorch_lightning/_graveyard/cloud_io.py (hunk 2 above)
Comment (justusschock) [reply src/pytorch_lightning/_graveyard/cloud_io.py:18]: They cannot do much to prevent this. The issue is that we did pickle entire classes in old checkpoints which results in pytorch_lightning being imported (or failure to load the checkpoint at all if PL is not present standalone). This warning is just to be on the safe side, that users will be informed what we are doing. They also should only experience this while loading legacy checkpoints

**Comment by justusschock** [comment]
@awaelchli I would like to wait for @carmocca to review as I initially discussed this with him

**Comment by Borda** [comment]
> @awaelchli I would like to wait for @carmocca to review as I initially discussed this with him

just for the record, if this works correctly we can revert 0fd1be6baef34c011357e02f170f26e78dbe87ef

File: src/pytorch_lightning/_graveyard/cloud_io.py [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+from typing import Any
+
+
+class RedirectingUnpickler(pickle.Unpickler):
+    def find_class(self, module: str, name: str) -> Any:
+        if module.startswith("pytorch_" + "lightning"):
+            # for the standalone package this won't do anything,
+            # for the unified mirror package it will redirect the imports
+
+            old_module = deepcopy(module)
```
Comment (carmocca) [inline src/pytorch_lightning/_graveyard/cloud_io.py:13]: Why are we deepcopying if it's a plain string?

File: src/pytorch_lightning/_graveyard/cloud_io.py (hunk 2 above)
Comment (carmocca) [inline src/pytorch_lightning/_graveyard/cloud_io.py]: Isn't the current way printing the same twice?

```suggestion
                warnings.warn(f"Redirecting import of {old_module}.{name} to {module}.{name}")
```

File: src/pytorch_lightning/_graveyard/cloud_io.py [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+import warnings
+from copy import deepcopy
+from typing import Any
+
+
+class RedirectingUnpickler(pickle.Unpickler):
+    def find_class(self, module: str, name: str) -> Any:
+        if module.startswith("pytorch_" + "lightning"):
+            # for the standalone package this won't do anything,
+            # for the unified mirror package it will redirect the imports
```
Comment (carmocca) [inline src/pytorch_lightning/_graveyard/cloud_io.py:12]: Have you considered cases where the user actually has both installed? Would there be any implications from this change?

Perhaps we should also not run this code if both are available

File: src/pytorch_lightning/_graveyard/__init__.py [hunk 5]
```diff
@@ -13,6 +13,7 @@
 # limitations under the License.
 
 import pytorch_lightning._graveyard.callbacks
+import pytorch_lightning._graveyard.cloud_io
```
Comment (carmocca) [inline src/pytorch_lightning/_graveyard/__init__.py:16]: Why did you choose `cloud_io`? This has nothing todo with our `src/lightning_lite/utilities/cloud_io.py` module, no?

I would suggest
```suggestion
import pytorch_lightning._graveyard.legacy_import_unpickler
```

Or something along those lines

(remember the test name too)

File: src/pytorch_lightning/_graveyard/cloud_io.py [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+            module = "pytorch_lightning" + module[len("pytorch_" + "lightning") :]
+
+            # this warning won't trigger for standalone as these imports are identical
+            if module != old_module:
+                warnings.warn(f"Redirecting import of {module}.{name} to pytorch_lightning.{name}")
+
+        return super().find_class(module, name)
+
+
+pickle.Unpickler = RedirectingUnpickler  # type: ignore
```
Comment (carmocca) [inline src/pytorch_lightning/_graveyard/cloud_io.py:21]: (not for this PR) I'm not sure the graveyard is the right place for this code.

Could it be added as a migration piece in the future? ref: #15237 

If we did, then maybe we can restore this `pickle.Unpickler` override back to the original

**Review by carmocca — APPROVED** [review]

File: src/pytorch_lightning/_graveyard/__init__.py (hunk 5 above)
Comment (justusschock) [reply src/pytorch_lightning/_graveyard/__init__.py:16]: Fine with me as well. I chose `cloud_io` since our basic load function is within the `cloud_io` package.

File: src/pytorch_lightning/_graveyard/cloud_io.py (hunk 4 above)
Comment (justusschock) [reply src/pytorch_lightning/_graveyard/cloud_io.py:12]: I haven't considered this case so far. But in that case it would depend which package is used to load the checkpoint. I.e when loaded with `lightning.pytorch` I'd argue that this should be executed but if you load with `pytorch_lightning` it should not. And this is also the way it behaves.

File: src/pytorch_lightning/_graveyard/cloud_io.py (hunk 3 above)
Comment (justusschock) [reply src/pytorch_lightning/_graveyard/cloud_io.py:13]: This is a leftover, where `module` was changed in place. Should be safe to delete now.

File: src/pytorch_lightning/_graveyard/cloud_io.py (hunk 6 above)
Comment (justusschock) [reply src/pytorch_lightning/_graveyard/cloud_io.py:21]: Exactly my thoughts. I didn't want to interfer with existing checkpoint PRs however. I chose graveyard because it is for legacy checkpoints (we don't support pickling anymore) and graveyard was where we had patching of modules anyways.

I'd prefer to move it to checkpointing after #15237 .

**Review by otaj — APPROVED** [review]
This is awesome, @justusschock, thanks for working on this!

File: tests/tests_pytorch/graveyard/test_legacy_import_unpickler.py [hunk 7]
```diff
@@ -0,0 +1,66 @@
+import glob
+import os
+import sys
+
+import pytest
+import torch
+
+import pytorch_lightning  # noqa: F401
+from tests.tests_pytorch.helpers.utils import no_warning_call
```
Comment (otaj) [inline tests/tests_pytorch/graveyard/test_legacy_import_unpickler.py:9]:
```suggestion
from tests_pytorch.helpers.utils import no_warning_call
```

File: tests/tests_pytorch/graveyard/test_legacy_import_unpickler.py [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+def test_imports_standalone(pl_version: str):
+    assert any(
+        key.startswith("pytorch_" + "lightning") for key in sys.modules.keys()
+    ), "Imported PL, so it has to be in sys.modules"
+    path_legacy = os.path.join(LEGACY_CHECKPOINTS_PATH, pl_version)
+    path_ckpts = sorted(glob.glob(os.path.join(path_legacy, f"*{CHECKPOINT_EXTENSION}")))
+    assert path_ckpts, f'No checkpoints found in folder "{path_legacy}"'
+    path_ckpt = path_ckpts[-1]
+
+    with no_warning_call():
```
Comment (akihironitta) [inline tests/tests_pytorch/graveyard/test_legacy_import_unpickler.py:30]: It's better to specify which warning exactly?
```suggestion
    with no_warning_call(match="Redirecting import of*"):
```

**Review by akihironitta — APPROVED** [review]

**Review by tchaton — APPROVED** [review]
LGTM !

**Comment by carmocca** [comment]
Bad news. This is failing in #15105: https://github.com/Lightning-AI/lightning/actions/runs/3328686095/jobs/5504966109

**Comment by justusschock** [comment]
@carmocca I'm on it!
